shared/js/advanced_telemetry_helper.js

```diff
@@ -0,0 +1,190 @@
+/*
+ * This module is a helper for recording telemetry data in Gecko’s telemetry
+ * system. Currently, only data from certified apps will be recorded.
+ *
+ * There are three different types of metrics that can be used here:
+ * Counter:  This is a simple counter that increments by 1 each time it's
+ *           called.
+ *
+ * Linear:   This type of metric creates a linear distributed set of buckets
+ *           which range between the min and max.  You can specify the number
+ *           of buckets as well as the min and max.
+ *
+ * Exponential: This type of metric creates an exponential distributed set of
+ *              buckets ranging between min and max.  You can specify the min,
+ *              max, and buckets.
+ *
+ * App Setup:
+ *   1) Add the following to the app (e.g. index.html):
+ *   <script defer src="/shared/js/advanced_telemetry_helper.js"></script>
+ *   <script defer src="/shared/js/settings_listener.js"></script>
+ *   2) Include AdvancedTelemetryHelper in the globals.
+ *
+ * Counter Metric Usage Example:
+ *  var countMetric = new AdvancedTelemetryHelper();
+ *  // This creates and increments the counter on this metric called
+ *  // 'mycountmetric'
+ *  countMetric.count('mycountmetric'); // Histogram is created
+ *  countMetric.log(); //The count of the histogram is now 1.
+ *  // DO SOMETHING
+ *  countMetric.log(); //The count of the histogram is now 2.
+ *
+ * Linear Metric Usage Example:
+ *  var linearMetric = new AdvancedTelemetryHelper();
+ *  // This creates and logs a linear histogram called 'mylinearmetric' with
+ *  // minimum of 1 (0 not allowed as a minimum), max of 1000, with 10 buckets.
+ *  linearMetric.linear('mylinearmetric', 1, 1000, 10);
+ *  // DO SOMETHING
+ *  // This adds to the existing histogram a value of 800.
+ *  linearmetric.log(800); // The histogram now has a value in the bucket
+ *                            // that holds the value 800
+ *
+ *  Exponential Metric Usage Example:
+ *  var expMetric = new AdvancedTelemetryHelper();
+ *  // This creates and logs a linear histogram called 'myexpmetric' with
+ *  // minimum of 1 (0 not allowed as a minimum), max of 1000, with 10 buckets.
+ *  expMetric.exp('myexpmetric', 1, 1000, 10);
+ *  // DO SOMETHING
+ *  // This adds to the existing histogram a value of 800.
+ *  expMetric.log(800); // The histogram now has a value in the bucket
+ *                      // that holds the value 800
+ *
+ */
+
+/* globals SettingsListener */
+(function(exports) {
+  'use strict';
+
+  const APP_TELEMETRY_LOG_PREFIX = 'telemetry';
+  const ATH = AdvancedTelemetryHelper;
+  const HISTOGRAM_COUNT = 4;
+  const HISTOGRAM_EXP = 0;
+  const HISTOGRAM_LINEAR = 1;
+  // This is exposed for unit testing purposes
+  ATH.TELEMETRY_ENABLED_KEY = 'debug.performance_data.advanced_telemetry';
+  ATH.SETTINGS_LISTENER = false;
```

Comment:
This variable name could be more descriptive. As it is, it's purpose is not obvious.